### ISSUE TITLE
Current master fails fork test with OpenSSL-1.1.1

### ISSUE THREAD

**Issue by mouse07410** [OpenSC/libp11#217]
MacOS 10.13.4, Xcode-9.3, current master OpenSSL (`OpenSSL 1.1.1-pre5-dev  xx XXX xxxx`). Problem seems to be with generating random numbers on a token:

```
$ ./configure --disable-silent-rules --prefix=/Users/uri/openssl-1.1 --with-pkcs11-module=/opt/local/lib/p11-kit-proxy.dylib --with-enginesdir=/Users/uri/openssl-1.1/lib/engines-1.1/
. . . . .
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: rsa-testpkcs11.softhsm
PASS: rsa-testfork.softhsm
PASS: rsa-testlistkeys.softhsm
PASS: rsa-evp-sign.softhsm
PASS: ec-testfork.softhsm
FAIL: fork-change-slot.softhsm
============================================================================
Testsuite summary for libp11 0.4.8_git
============================================================================
# TOTAL: 6
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
============================================================================
See tests/test-suite.log
============================================================================
make[3]: *** [test-suite.log] Error 1
make[2]: *** [check-TESTS] Error 2
make[1]: *** [check-am] Error 2
make: *** [check-recursive] Error 1
$ cat tests/test-suite.log 
============================================
   libp11 0.4.8_git: tests/test-suite.log
============================================

# TOTAL: 6
# PASS:  5
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: fork-change-slot.softhsm
==============================

Current directory: /Users/uri/src/libp11/tests
Source directory: .
Output directory: output.76528
-n * Initializing smart card... 
ok
-n * Initializing smart card... 
ok
Key pair generated:
Private Key Object; RSA 
  label:      pkey
  ID:         01020304
  Usage:      decrypt, sign, unwrap
Public Key Object; RSA 1024 bits
  label:      pkey
  ID:         01020304
  Usage:      encrypt, verify, wrap
Found token (39dc8410-4c78-0dbe-dd48-bbf84b72866b) with matching token label.
The token (output.76528/softhsm-testpkcs11.db/39dc8410-4c78-0dbe-dd48-bbf84b72866b) has been deleted.
pid 76553: RAND_bytes generated errors:
140735502353280:error:82065005:PKCS#11 module:pkcs11_check_token:General Error:p11_slot.c:487:
pid 76538 is the parent
Remove or add a device to try to cause an error
Waiting for signal SIGUSR1
pid 76538 is the parent
spawned 76553
child 76553 exited with status 1
FAIL fork-change-slot.softhsm (exit status: 1)
```

This used to work before #214. Need to mention that SoftHSMv2 was patched, otherwise there's infinite recursion (see #215).

Here's the SoftHSMv2 patch: 
```
diff --git a/src/lib/crypto/OSSLCryptoFactory.cpp b/src/lib/crypto/OSSLCryptoFactory.cpp
index 8555815..ae77157 100644
--- a/src/lib/crypto/OSSLCryptoFactory.cpp
+++ b/src/lib/crypto/OSSLCryptoFactory.cpp
@@ -141,6 +141,25 @@ OSSLCryptoFactory::OSSLCryptoFactory()
 	// Initialise OpenSSL
 	OpenSSL_add_all_algorithms();
 
+	// Make sure RDRAND is loaded first
+	ENGINE_load_rdrand();
+	// Locate loaded RDRAND engine
+	rdrand_engine = ENGINE_by_id("rdrand");
+	if ( rdrand_engine == NULL ) {
+		fprintf(stderr, "ENGINE_load_rdrand returned %lu\n", ERR_get_error());
+	}
+	// Initialize RDRAND engine
+	if ( ! ENGINE_init(rdrand_engine) ) {
+		fprintf(stderr, "ENGINE_init returned %lu\n", ERR_get_error());
+	}
+	// Set RDRAND engine as default for RAND_ - this is necessary for
+	// pkcs11 engines that also provide random number generators.
+	// Without this SoftHSMv2 goes into infinite recursion trying to
+	// get random numbers from an engine that's being initialized.
+	if ( ! ENGINE_set_default(rdrand_engine, ENGINE_METHOD_RAND) ) {
+		fprintf(stderr, "ENGINE_set_default returned %lu\n", ERR_get_error());
+	}
+
 	// Initialise the one-and-only RNG
 	rng = new OSSLRNG();
 
diff --git a/src/lib/crypto/OSSLCryptoFactory.h b/src/lib/crypto/OSSLCryptoFactory.h
index e8bfa2c..8dfb128 100644
--- a/src/lib/crypto/OSSLCryptoFactory.h
+++ b/src/lib/crypto/OSSLCryptoFactory.h
@@ -105,6 +105,8 @@ private:
 
 	// The one-and-only RNG instance
 	RNG* rng;
+	// And RDRAND engine to use with it
+	ENGINE *rdrand_engine;
 
 #ifdef WITH_GOST
 	// The GOST engine
```

**Comment by mouse07410** [comment]
It is worth mentioning that it's the `RAND_bytes()` that causes this error. When I comment out the following lines https://github.com/OpenSC/libp11/blob/4894744f9b66c956ddef2cbd232ddeb1741205be/tests/fork-change-slot.c#L228-L232, all the tests pass fine.

So there is something with how this new RAND method is implemented, as, e.g., I've no problem getting randomness from `RDRAND_engine`.

**Comment by williamcroberts** [comment]
@mouse07410 
The reason you openssl command was returning NO random values is that [tok->initialized](https://github.com/OpenSC/libp11/blob/master/src/eng_back.c#L1028)  is false, and thus nothing is available for rng purposes.
Also note that even if you coerce the values in a debugger to get all the to the C_GetRandom call in SoftHSM2, when it calls RAND_bytes() it calls back to the pkcs#11 engine, which then calls C_GetRandom... so you'll end up in another recursion. SoftHSMV2 would need to be smart enough to back-up, switch randrd and restore the pkcs11engine.

**Comment by mouse07410** [comment]
> The reason you openssl command was returning NO random values is that `tok->initialized` is false, and thus nothing is available for rng purposes.

OK, I understand. What's the answer then?

> SoftHSMV2 would need to be smart enough to back-up, switch randrd and restore the pkcs11engine.

That sounds rather complicated. Is it something that libp11 can address on its own (create a workaround), or should it be brought as a patch to SoftHSMv2? If latter - are you willing to submit a PR there?

**Comment by williamcroberts** [comment]
For OSSL 1.0.2, I have no idea. With SoftHSMv2 and Openssl 1.0.2g I get no output. I only get output with my hardware module.

For OSSL 1.1, I have no idea.

Their is nothing to fix with SoftHSMv2 and it's integration with pkcs11 engine. This is a limitation in that the openssl design never accounted for and is only an issue when the engine code calls openssl code that invokes an engine.

**Comment by mouse07410** [comment]
> With SoftHSMv2 and Openssl 1.0.2g I get no output. I only get output with my hardware module.

Ah. With SoftHSMv2 I only care for the tests to pass. ;-)
And indeed I do get output from my hardware module.

> For OSSL 1.1, I have no idea.

But it would be nice to understand why with OSSL 1.0.2[g-o] you can get output from hardware module, and with OSSL 1.1.x you cannot.

Also, even when I comment the [tok->initialized](https://github.com/OpenSC/libp11/blob/4894744f9b66c956ddef2cbd232ddeb1741205be/src/eng_back.c#L1028) line out, OSSL 1.1 still does not provide any output.

> This is a limitation in that the openssl design never accounted for and is only an issue when the engine code calls openssl code that invokes an engine.

Well, OSSL design has been around for a long time. And I for one am not happy about a modification to an OSSL-related package that breaks with the current OSSL release.

This recursion problem does not appear in other methods. Shouldn't there be a way to prevent it here? I suspect the majority on the OSSL team would say "it works fine without this libp11 change - figure how to make it work inside libp11". 

What are we going to do? OSSL 1.1 is the way, so libp11 must work with it, and at the very least not fail its own tests...

**Comment by mouse07410** [comment]
@williamcroberts @mtrojnar here's what I suggest we do: modify #214 to make the added code compile conditionally on OpenSSL <= 1.0.2 only. Because there's nothing but trouble with it on 1.1.x, and even the "normal" libp11 tests fail.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Friday, April 6, 2018 9:11 AM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Comment
> <comment@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> 	With SoftHSMv2 and Openssl 1.0.2g I get no output. I only get output
> with my hardware module.
> 
> Ah. With SoftHSMv2 I only care for the tests to pass. ;-) And indeed I do get
> output from my hardware module.
> 
> 	For OSSL 1.1, I have no idea.
> 
> But it would be nice to understand why with OSSL 1.0.2[g-o] you can get output
> from hardware module, and with OSSL 1.1.x you cannot.
> 
> Also, even when I comment the tok->initialized
> <https://github.com/OpenSC/libp11/blob/4894744f9b66c956ddef2cbd232ddeb17
> 41205be/src/eng_back.c#L1028>  line out, OSSL 1.1 still does not provide any
> output.
> 
> 	This is a limitation in that the openssl design never accounted for and is
> only an issue when the engine code calls openssl code that invokes an engine.
> 
> Well, OSSL design has been around for a long time. And I for one am not happy
> about a modification to an OSSL-related package that breaks with the current
> OSSL release.

As far as 1.1.x this is likely a bug un OSSL. Feel free to go find it.

> 
> This recursion problem does not appear in other methods. Shouldn't there be a
> way to prevent it here? I suspect the majority on the OSSL team would say "it
> works fine without this libp11 change - figure how to make it work inside libp11".

Because, for the moment, the calls into OSSL performed by SoftHSMv2 and the engines
registered by libp11 have never conflicted.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Monday, April 9, 2018 9:24 AM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> 	As far as 1.1.x this is likely a bug un OSSL. Feel free to go find it
> 
> Unfortunately, my time - just like yours - is limited now. And this problem is
> manifested by your PR.
> 
> 	Because, for the moment, the calls into OSSL performed by SoftHSMv2
> and the engines
> 	registered by libp11 have never conflicted.
> 
> OK, but they do conflict now - and libp11 fails to pass its own tests with OpenSSL-
> 1.1 (which is the mainstream now).

I'll look at 1.1 at some point soon. Probably a week or so.

Rather than checking OSSL version, which means as we find/fix OSSL bugs
We need to keep if defing and patching our code, I propose we make this
dynamic and disabled by default. If we choose too, we can set the default
based on OSSL version at configure time. When this stabilizes we can change
the default behavior. This also provides a way to override an already
configured package.

See PR: https://github.com/OpenSC/libp11/pull/222

**Comment by mouse07410** [comment]
There's no reason to disable it for versions that we know for sure work OK. And there's all the reason to not compile it for versions that we know for sure it fails (until the cause is found and the fix is implemented). I think this should be sufficient for now:
```
diff --git a/src/eng_front.c b/src/eng_front.c
index ec3d04f..bd11f39 100644
--- a/src/eng_front.c
+++ b/src/eng_front.c
@@ -241,7 +241,9 @@ static int bind_helper(ENGINE *e)
                        !ENGINE_set_ctrl_function(e, engine_ctrl) ||
                        !ENGINE_set_cmd_defns(e, engine_cmd_defns) ||
                        !ENGINE_set_name(e, PKCS11_ENGINE_NAME) ||
+#if OPENSSL_VERSION_NUMBER  < 0x10100002L
                        !ENGINE_set_RAND(e, PKCS11_get_rand_method()) ||
+#endif /* OPENSSL_VERSION_NUMBER */
 #ifndef OPENSSL_NO_RSA
                        !ENGINE_set_RSA(e, PKCS11_get_rsa_method()) ||
 #endif
@@ -263,6 +265,9 @@ static int bind_helper(ENGINE *e)
                        !ENGINE_set_load_privkey_function(e, load_privkey)) {
                return 0;
        } else {
+#if OPENSSL_VERSION_NUMBER  < 0x10100002L
+               ENGINE_set_table_flags(ENGINE_TABLE_FLAG_NOINIT);
+#endif /* OPENSSL_VERSION_NUMBER */
                ERR_load_ENG_strings();
                return 1;
        }
```

And indeed, when this all stabilizes, we can change it to whatever the OSSL-1.1 fix may require. I suspect there would have to be two versions of `PKCS11_RAND_...` code (the current one that works for 1.0.2, and something different to accommodate whatever OSSL-1.1 would do).

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Monday, April 9, 2018 11:35 AM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> There's no reason to disable it for versions that we know for sure work OK. And
> there's all the reason to not compile it for versions that we know for sure it fails
> (until the cause is found and the fix is implemented). I think this should be
> sufficient for now:
> 
> diff --git a/src/eng_front.c b/src/eng_front.c index ec3d04f..bd11f39 100644
> --- a/src/eng_front.c
> +++ b/src/eng_front.c
> @@ -241,7 +241,9 @@ static int bind_helper(ENGINE *e)
>                         !ENGINE_set_ctrl_function(e, engine_ctrl) ||
>                         !ENGINE_set_cmd_defns(e, engine_cmd_defns) ||
>                         !ENGINE_set_name(e, PKCS11_ENGINE_NAME) ||
> +#if OPENSSL_VERSION_NUMBER  < 0x10100002L
>                         !ENGINE_set_RAND(e, PKCS11_get_rand_method()) ||

We don't know if this is needed in all versions less than this....

> +#endif /* OPENSSL_VERSION_NUMBER */
>  #ifndef OPENSSL_NO_RSA
>                         !ENGINE_set_RSA(e, PKCS11_get_rsa_method()) ||  #endif @@ -
> 263,6 +265,9 @@ static int bind_helper(ENGINE *e)
>                         !ENGINE_set_load_privkey_function(e, load_privkey)) {
>                 return 0;
>         } else {
> +#if OPENSSL_VERSION_NUMBER  < 0x10100002L

We don't know if this is needed in all versions less than this....

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Monday, April 9, 2018 11:35 AM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> There's no reason to disable it for versions that we know for sure work OK. And
> there's all the reason to not compile it for versions that we know for sure it fails
> (until the cause is found and the fix is implemented). I think this should be
> sufficient for now:

Sufficient yes, but clients will have no idea if their version of pkcs#11 engine supports
random without knowing about this patch and the version it was configured and linked
against.

I'll update my dynamic PR with this PR, this way it makes sense.
1. patch to fix the recursive problem
2. patch to make it configurable dynamic
3. patch in the testing code to enable it since it works on travis.

**Comment by williamcroberts** [comment]
Item 3 above gets harder as https://github.com/OpenSC/libp11/blob/master/tests/evp-sign.c#L178 fails if you try and use a config option that doesn't take an argument. Even something like FORCE_LOGIN will fail to load.

**Comment by mouse07410** [comment]
> Sufficient yes, but clients will have no idea if their version of pkcs#11 engine supports
random without knowing about this patch and the version it was configured and linked
against.

Any client can safely assume that this version of pkcs#11 engine does with with OSSL-1.0.2, as has been tested on various platforms. No need for the client to "have an idea", and no reason to ever disable this feature (don't need it - just don't use it, nothing will break).

Any client should assume that this version of the engine will **not** work with any OSSL-1.1.x. Again, no "idea" would help, and no amount of dynamic configuration is going to change that.

In other words, this version of pkcs#11 engine does work with 1.0.2 so no dynamic configuration is necessary, and it does not work with 1.1.x so no dynamic configuration can help.

And of course 1.0.2 and 1.1.x cannot share the engine binaries - so no mix/confusion is possible here.

Also, there's an important issue: how does an app use the dynamic configuration interface? Are you going to propose a patch to OpenSSL CLI? I strongly suspect it would be rejected, even if you do... 

The more I think about it, the more I think that it's a bad idea.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Monday, April 9, 2018 4:33 PM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> 	Sufficient yes, but clients will have no idea if their version of pkcs#11
> engine supports
> 	random without knowing about this patch and the version it was
> configured and linked
> 	against.
> 
> Any client can safely assume that this version of pkcs#11 engine does with with
> OSSL-1.0.2, as has been tested on various platforms. No need for the client to
> "have an idea", and no reason to ever disable this feature (don't need it - just
> don't use it, nothing will break).
> 
> Any client should assume that this version of the engine will not work with any
> OSSL-1.1.x. Again, no "idea" would help, and no amount of dynamic configuration
> is going to change that.
> 
> In other words, this version of pkcs#11 engine does work with 1.0.2 so no
> dynamic configuration is necessary, and it does not work with 1.1.x so no dynamic
> configuration can help.

What if the bug is fixed on 1.1.2, broken on 1.1.1 and 1.1.3? This is where if
Defing the code becomes a PITA maintenance wise and the reason we don't
Do it this way in the Linux kernel. If we see instabilities, we default to know,
mark it as experimental until we know all the issues and then swap the
default value.

**Comment by mouse07410** [comment]
This seems a "what if" conversation vs. "what is". Can you name a couple of examples of things that were fixed in OSSL 1.1.x but got broken in 1.1.(x-1) and 1.1.(x+1)?

> Def'ing the code becomes a PITA maintenance wise

When did you look at the libp11 code last? Noticed OPENSSL_VERSION ifdefs?

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Mouse [mailto:notifications@github.com]
> Sent: Tuesday, April 10, 2018 5:01 PM
> To: OpenSC/libp11 <libp11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [OpenSC/libp11] Current master fails fork test with OpenSSL-1.1.1
> (#217)
> 
> This seems a "what if" conversation vs. "what is". Can you name a couple of
> examples of things that were fixed in OSSL 1.1.x but got broken in 1.1.(x-1) and
> 1.1.(x+1)?

It's called a regression, and they happen all the time. Here's some:
https://github.com/openssl/openssl/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+regression
and those where the only ones people mentioned that keyword in. Likely more.

> 
> 	Def'ing the code becomes a PITA maintenance wise
> 
> When did you look at the libp11 code last? Noticed OPENSSL_VERSION ifdefs?

Their working around issues/api changes but providing stable support to the user.
For instance in this (which you even suggested):
https://github.com/OpenSC/libp11/blob/master/src/eng_front.c#L205

A few ifdef's are normal, but you should add them sparingly. In the cases there,
they make sense.

**Comment by mouse07410** [comment]
> It's called a regression, and they happen all the time. Here's some:...

When a regression happens, it gets fixed. In other words, I don't think it is reasonable to design code with an assumption that there would be a regression that would occasionally break it, and would not be fixed in the given version.
